### ISSUE TITLE
heading-order scope & related nodes bugs

### ISSUE THREAD

**Issue by WilcoFiers** [dequelabs/axe-core#728]
The heading-order rule makes the assumption that it's always used to test the entire page. That's incorrect, because the context object can have it test for instance only the header and footer. This can introduce bugs in two ways: 
1. An "in-between" heading is overlooked, creating a false positive (see below)
2. Issues can be overlooked when they cross the scope boundary.

Here's a use case:
```html
<!doctype html>
<html lang="en">
<title>O hai</title>
<header>
  <h1>Foo</h1>
</header>
<main>
  <h2>bar</h2>
</main>
<footer>
  <h3>Baz</h3>
</footer>

<script src="/axe.js"></script>
<script>
window.addEventListener('load' , function () {
  axe.run({ include: [['header'], ['footer']] }, {
      runOnly: { type: "rule", values: ["heading-order"] }
    }, function (err, res) {
      // Expect: Undefined
      // Actual: heading-order violation
      console.log(res.violations[0]);
    });
})
</script>
```

To solve this, instead of using an 'after' method, we should look for the *previous* heading on the page.

### Minor bug
I also noticed that the previous heading wasn't included as a related node. We should fix that.

**Comment by marcysutton** [comment]
Good timing. I was about to make changes to this rule for Shadow DOM, but I hadn't fully looked into it yet. There are a number of checks in `navigation` that need to be updated–I've gotten up to `p-as-heading` so far. I'll take a look at this one next week.

**Comment by marcysutton** [comment]
This appears to be more complicated than I first thought, and it warrants a discussion about how it should work.

**Comment by WilcoFiers** [comment]
What's the concern Marcy?

**Comment by dylanb** [comment]
Why not simply flag this as a page level rule?

**Comment by WilcoFiers** [comment]
Page level rules can still have parts of the page scoped out. That doesn't solve the issue.

**Comment by dylanb** [comment]
@WilcoFiers that is a bit of a conundrum...perhaps we should add an option for page level rules to be turned off on these scenarios? Seems like there is no really good behavior that works for all scenarios.

**Comment by WilcoFiers** [comment]
I still don't see why we couldn't just have this rule crawl backward up the DOM and look for the previous visible heading. Seems like that would solve this pretty easily.

**Comment by dylanb** [comment]
Because that does not work across iframe boundaries

**Comment by padmavemulapati** [comment]
Verified with the latest axe-core-develop branch code (dated 10/12/2020) -
As heading-order overlooked unscoped and in-between headings - different set of cases constructed and validated.
Below are the cases:
1. If heading-order with frames
code snippet: `<head> <meta charset="utf8" /> <script src="/axe.js"></script> </head>`
	`<body> <h3>Heading 3</h3> <iframe src="nested-heading.html"></iframe> </body> `
![image](https://user-images.githubusercontent.com/47409579/101776577-279fc200-3b17-11eb-83aa-95898f8f7aa4.png)

2. Nested headings:
code snippet: `<head> <meta charset="utf8" /> <script src="/axe.js"></script> </head> `
	`<body> <h4>Heading 4</h4> </body> `
![image](https://user-images.githubusercontent.com/47409579/101776923-aeed3580-3b17-11eb-9faf-438d85472cb6.png)

3. Partial-context-with-iframe-excluded:
code snippet: `<iframe src="frames/heading.html"></iframe>`
		`<footer> <h4>Baz</h4> </footer>`
![image](https://user-images.githubusercontent.com/47409579/101777168-186d4400-3b18-11eb-88a5-f2c4b7e5c7d0.png)

4. Partial-order-context-with-iframes:
code snippet: `<header> <h1>Foo</h1> </header> `    ` <main> <h2>bar</h2> </main>`
			`<iframe src="frames/heading.html"></iframe> `
		 `<footer> <h5>Baz</h5> </footer> `
![image](https://user-images.githubusercontent.com/47409579/101777417-769a2700-3b18-11eb-9b88-f5a211fd5714.png)

5. with partial-context:
code snippet: `<header> <h1>Foo</h1> </header>`  `<main> <h2>bar</h2> </main>` `<footer> <h3>Baz</h3> </footer>`
![image](https://user-images.githubusercontent.com/47409579/101777662-d395dd00-3b18-11eb-890c-8cd53ac9c90d.png)
